### PR TITLE
Fixed upper bound buffer length in tok_key_get_key_type() and add remove-key command

### PR DESCRIPTION
**Pull request by reumann-ibm**

### CODE REVIEW

File: usr/sbin/p11sak/p11sak.c [hunk 1]
```diff
@@ -1356,6 +1365,8 @@ static CK_RV tok_key_get_key_type(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE hk
                 ret);
         return ret;
     }
+
+
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:1369]: did you add these 2 blank lines by intention? 1 blank line should be sufficient.....

Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:1369]: No intention. Removed one blank line.

File: usr/sbin/p11sak/p11sak.c [hunk 2]
```diff
@@ -1356,6 +1368,7 @@ static CK_RV tok_key_get_key_type(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE hk
                 ret);
         return ret;
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c]: I think buffer also needs to be free'd here (probably at all error exits). Think about providing an error exit label for this function.

File: usr/sbin/p11sak/p11sak.c [hunk 3]
```diff
@@ -844,13 +865,13 @@ static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_ULONG keylength,
...
-                    CKA_VERIFY, &a_verify, bs }, { CKA_WRAP, &a_wrap, bs }, {
-                    CKA_UNWRAP, &a_unwrap, bs }, { CKA_EXTRACTABLE,
-                    &a_extractable, bs },
+            CKA_MODIFIABLE, &a_modifiable, bs }, { CKA_DERIVE, &a_derive, bs },
+            { CKA_SENSITIVE, &a_sensitive, bs }, {
+            CKA_ENCRYPT, &a_encrypt, bs }, { CKA_DECRYPT, &a_decrypt, bs }, {
+                    CKA_SIGN, &a_sign, bs }, {
+            CKA_VERIFY, &a_verify, bs }, { CKA_WRAP, &a_wrap, bs }, {
+            CKA_UNWRAP, &a_unwrap, bs },
+            { CKA_EXTRACTABLE, &a_extractable, bs },
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:874]: There a few chunks in this commit for which its hard to tell if you actually changed something. E.g., this seems to be formatting only. If you think the code needs re-formatting, please make it a separate commit next time.

Comment (p-steuer) [reply usr/sbin/p11sak/p11sak.c:874]: (The same applies to renaming variables or, more general, all changes that are unrelated to the commit message)

File: usr/sbin/p11sak/p11sak.c [hunk 4]
```diff
@@ -1910,104 +2051,264 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+
+    rc = tok_key_list_init(session, kt, label);
+    if (rc != CKR_OK) {
+        printf("Init token key list failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
+        return rc;
+    }
+
+    while (1) {
+        rc = funcs->C_FindObjects(session, &hkey, 1, &count);
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2092]: You only get 1 find result per loop iteration. It works that way, but it would perform much better if you get multiple (say 10 or 20) results with one call to C_FindObjects and the have a secondary loop to iterate over the results (0 .. count -1).

File: usr/sbin/p11sak/p11sak.c [hunk 5]
```diff
@@ -1910,104 +2051,264 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+                    printf("Are you sure to delete %s? (y/n)\n", label);
+
+                    nread = getline(&buff, &buflen, stdin);
+                    if (nread == -1) {
+                        rc = -1;
+                        goto done;
+                    }
+                    buff[nread - 1] = '\0';
+
+                    if (strcmp(buff, "y") == 0) {
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2177]: Some lines seem to be a duplicate from above.  Maybe this can be refactored to use a subroutine for the common parts?

File: usr/sbin/p11sak/p11sak.c [hunk 6]
```diff
@@ -1910,104 +2051,264 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+                            "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                            label, rc, p11_get_ckr(rc));
+                    return rc;
+                }
+                printf("DONE - Destroy Object with Label: %s\n", label);
+            } else {
+                while (1) {
+                    printf("Are you sure to delete %s? (y/n)\n", label);
+
+                    nread = getline(&buff, &buflen, stdin);
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2170]: I guess you are intentionally not freeing buff within the loop iteration, so that it is still allocated from the previous iteration here, and is reused/reallocted by getline ?

File: usr/sbin/p11sak/p11sak.c (hunk 5 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:2177]: I have introduced a couple of subroutines to refactore the code as suggested.

File: usr/sbin/p11sak/p11sak.c (hunk 3 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:874]: Ok. Will do. Thanks for the suggestion.

File: usr/sbin/p11sak/p11sak.c (hunk 2 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c]: Fixed. buffer and label need to be freed in the while(1) loop.

File: usr/sbin/p11sak/p11sak.c (hunk 4 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:2143]: This will become an item for future implementation.

**Comment by reumann-ibm** [comment]
The comments and suggestions are worked into the code and the commit has been amended.

File: usr/sbin/p11sak/p11sak.c [hunk 7]
```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+static CK_RV getKeyboardInput(char *buff, char *label)
+{
+
+    CK_RV rc = CKR_OK;
+    int nread;
+    size_t buflen;
+
+    printf("Are you sure to delete %s? (y/n)\n", label);
+
+    nread = getline(&buff, &buflen, stdin);
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2083]: I don't think that this works as intended. 
getline() allocates  buff, if buff is NULL. If buff is not NULL, it checks if buflen is large enough to hold the input. If it is not large enough then it will reallocate the buffer.
You do allocate buff in the calling routine, but only in 1 byte length:
`if (!(buff = malloc(strlen("y"))))`
Thats wrong in 3 places:
1. strlen("y") returns 1, but you also need space for the terminating zero.
2. buflen is not initialized, so getline will get garbage as buffer length.
3. What if the user input is larger than one character? Then getline will reallocate buff, but you don't pass it back to the caller.

File: usr/sbin/p11sak/p11sak.c [hunk 8]
```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+static CK_RV deleteKey(CK_SESSION_HANDLE session, p11sak_kt kt, char *rm_label,
+                       CK_BBOOL *forceAll)
+{
+    CK_ULONG keylength, count;
+    CK_OBJECT_CLASS keyclass;
+    CK_OBJECT_HANDLE hkey;
+    char *keytype = NULL;
+    char *label = NULL;
+    CK_RV rc;
+    char *buff = NULL;
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2128]: I would not call this 'buff', name it 'user_input', or anything that tells the reader what it is used for.

File: usr/sbin/p11sak/p11sak.c [hunk 9]
```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+                rc = funcs->C_DestroyObject(session, hkey);
+                if (rc != CKR_OK) {
+                    printf(
+                            "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                            label, rc, p11_get_ckr(rc));
+                    return rc;
+                }
+                printf("DONE - Destroy Object with Label: %s\n", label);
+            } else {
+                while (1) {
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2199]: what is this while loop for? You exit the loop either by return or by break in any case after one iteration.

File: usr/sbin/p11sak/p11sak.c [hunk 10]
```diff
@@ -1908,106 +2053,277 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+                    rc = getKeyboardInput(buff, label);
+                    if (rc != CKR_OK) {
+                        goto done;
+                    }
+
+                    rc = finalizeDestroyObject(buff, label, &session, &hkey);
+                    if (rc != CKR_OK) {
+                        return rc;
+                    }
+                    else {
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:2209]: else on the same line as the closing bracket

File: usr/sbin/p11sak/p11sak.c [hunk 11]
```diff
@@ -113,20 +114,57 @@ static CK_RV get_pin(char **pin, size_t *pinlen) {
...
+{
+    int nread;
+    char *keyboard_input = NULL;
+    size_t buflen;
+    CK_RV rc = 0;
+
+    /* read the keyboard_input
+     * Note: getline will allocate memory for keyboard_input. free it when done.
+     */
+    nread = getline(&keyboard_input, &buflen, stdin);
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:138]: Why not directly pass user_input to getline? 

*user_input  = NULL;
nread = getline(user_input , &buflen, stdin);

Then you do not need to allocate a new buffer and copy the content.

File: usr/sbin/p11sak/p11sak.c [hunk 12]
```diff
@@ -113,20 +114,57 @@ static CK_RV get_pin(char **pin, size_t *pinlen) {
...
+    if (nread == -1) {
+        rc = -1;
+        goto done;
+    }
+
+    /* Allocate keyboard_input.
```
Comment (ifranzki) [inline usr/sbin/p11sak/p11sak.c:144]: See above.
If you really need to duplicate the buffer, why not simply use strdup() ?

The buffer allocated by getline is supposed to be zero terminated, so stdup should just work fine.

**Comment by ifranzki** [comment]
Travis build errors in https://travis-ci.org/github/opencryptoki/opencryptoki/builds/721927834:

In file included from /usr/include/string.h:494:0,
                 from usr/sbin/p11sak/p11sak.c:16:
In function ‘strcpy’,
    **inlined from ‘deleteKey’ at usr/sbin/p11sak/p11sak.c:2252:21,
    inlined from ‘execute_cmd’ at usr/sbin/p11sak/p11sak.c:2332:12,
    inlined from ‘main’ at usr/sbin/p11sak/p11sak.c:2481:8:**
/usr/include/s390x-linux-gnu/bits/string_fortified.h:90:10: error: ‘__builtin___memcpy_chk’ writing 2 bytes into a region of size 1 overflows the destination [-Werror=stringop-overflow=]
   return __builtin___strcpy_chk (__dest, __src, __bos (__dest));
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

usr/sbin/p11sak/p11sak.c:1858:17: error: equality comparison with extraneous parentheses [-Werror,-Wparentheses-equality]
    if ((*label == NULL)) {
         ~~~~~~~^~~~~~~
usr/sbin/p11sak/p11sak.c:1858:17: note: remove extraneous parentheses around the comparison to silence this warning
    if ((*label == NULL)) {
        ~       ^      ~
usr/sbin/p11sak/p11sak.c:1858:17: note: use '=' to turn this equality comparison into an assignment
    if ((*label == NULL)) {
                ^~
                =

File: usr/sbin/p11sak/p11sak.c (hunk 10 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c]: Changed as commented.

**Comment by ifranzki** [comment]
You now have 2 commits named "Added new command remove-key to the p11sak tool to remove key objects… " in the PR. Is this on intention? Shouldn't these 2 be squashed together?

File: usr/sbin/p11sak/p11sak.c (hunk 2 above)
Comment (p-steuer) [reply usr/sbin/p11sak/p11sak.c]: buffer still needs to be freed on below switch statement's error return.

File: usr/sbin/p11sak/p11sak.c (hunk 12 above)
Comment (p-steuer) [reply usr/sbin/p11sak/p11sak.c:144]: If your intention was to only modify the user_input arg in case of successful exit, you could just set user_input = keyboard_input in that case, instead of allocating and copying to a new buffer. Caller is responsible to free user_input any way.

File: usr/sbin/p11sak/p11sak.c (hunk 3 above)
Comment (p-steuer) [reply usr/sbin/p11sak/p11sak.c:939]: There is still a lot of formatting/renaming included in this commit. Thats okay for now, but please take care to put such changes in an extra commit (if its more than just a few lines) in future contributions.

File: usr/sbin/p11sak/p11sak.c [hunk 13]
```diff
@@ -1652,146 +1783,231 @@ static CK_RV parse_gen_key_args(char *argv[], int argc, p11sak_kt *kt,
...
+        } else if (strcmp(argv[i], "RSA") == 0 || strcmp(argv[i], "rsa") == 0) {
+            *kt = kt_RSAPKCS;
+        } else if (strcmp(argv[i], "EC") == 0 || strcmp(argv[i], "ec") == 0) {
+            *kt = kt_EC;
+            /* Get options */
+        } else if (strcmp(argv[i], "--slot") == 0) {
+            *slot = (CK_ULONG) atol(argv[i + 1]);
+            i++;
+        } else if (strcmp(argv[i], "--pin") == 0) {
+            *pin = argv[i + 1];
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:1835]: I think you have to check for an out of bounds read here: If "--poin" is the last string in argv, argv[i+1] becomes a NULL pointer deref.

File: usr/sbin/p11sak/p11sak.c [hunk 14]
```diff
@@ -1652,146 +1783,231 @@ static CK_RV parse_gen_key_args(char *argv[], int argc, p11sak_kt *kt,
...
+            i++;
+        } else if (strcmp(argv[i], "--label") == 0) {
+            *label = argv[i + 1];
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:1838]: Same here.

File: usr/sbin/p11sak/p11sak.c [hunk 15]
```diff
@@ -1908,109 +2129,266 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
-    free(keytype);
-    return ret;
+    return rc;
 }
 
-static CK_RV execute_cmd(CK_SESSION_HANDLE session, CK_SLOT_ID slot, p11sak_cmd cmd,
-        p11sak_kt kt, CK_ULONG keylength, CK_ULONG exponent, char *ECcurve,
-        char *label, char *attr_string, int long_print) {
-    CK_RV ret;
+static CK_RV confirmDestroy(char **user_input, char* label)
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:2150]: We usually dont use camel case for internal functions, do we ?

File: usr/sbin/p11sak/p11sak.c [hunk 16]
```diff
@@ -1908,109 +2129,266 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+                }
+                printf("DONE - Destroy Object with Label: %s\n", label);
+            }
+        } else {
+            if ((strcmp(rm_label, "") == 0) || (strcmp(rm_label, label) == 0)) {
+                rc = confirmDestroy(&user_input, label);
+                if (rc != CKR_OK) {
+                    printf("User input failed (error code 0x%lX: %s)\n",
+                            rc, p11_get_ckr(rc));
+                    return rc;
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:2263]: free(user_input) ?

File: usr/sbin/p11sak/p11sak.c [hunk 17]
```diff
@@ -1908,109 +2129,266 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+                }
+
+                rc = finalizeDestroyObject(user_input, label, &session, &hkey);
+                if (rc != CKR_OK) {
+                    return rc;
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:2268]: free(user_input) ?

File: usr/sbin/p11sak/p11sak.c [hunk 18]
```diff
@@ -1908,109 +2129,266 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
...
+    rc = funcs->C_FindObjectsFinal(session);
+    if (rc != CKR_OK) {
+        printf("C_FindObjectsFinal failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
+        goto done;
+    }
+
+    rc = CKR_OK;
+
+done:
```
Comment (p-steuer) [inline usr/sbin/p11sak/p11sak.c:2287]: Why not use this exit label for more error exits to free buffers etc ?

**Review by p-steuer — CHANGES_REQUESTED** [review]
Done with the review.

File: usr/sbin/p11sak/p11sak.c (hunk 18 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:2287]: Made use of the exit label to free buffers.

File: usr/sbin/p11sak/p11sak.c (hunk 17 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:2268]: Reworked user_input into finialize_destroy_object()

File: usr/sbin/p11sak/p11sak.c (hunk 16 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:2263]: Reworked user_input into finialize_destroy_object()

File: usr/sbin/p11sak/p11sak.c (hunk 15 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:2150]: Changed internal function names.

File: usr/sbin/p11sak/p11sak.c (hunk 14 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:1838]: Included out of bounds check.

File: usr/sbin/p11sak/p11sak.c (hunk 13 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:1835]: Included out of bounds check.

File: usr/sbin/p11sak/p11sak.c (hunk 12 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:144]: Simplified user input into confirm_destroy()

File: usr/sbin/p11sak/p11sak.c (hunk 11 above)
Comment (reumann-ibm) [reply usr/sbin/p11sak/p11sak.c:138]: Reworked user_input to remove allocating new buffers & content copy.

**Review by p-steuer — APPROVED** [review]
looks good
